## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
# import Pandas Libruary
import pandas as pd
# Define a formatting function to format values as currency. Only use for printing 
def format_currency(value):
    return '${:,.2f}'.format(value)

In [2]:
client_data_df = pd.read_csv('Resources/client_dataset.csv')
client_data_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [21]:
# View the column names in the data
client_data_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal', 'shipping_cost', 'shipping_weight',
       'line_total', 'line_cost', 'line_profit'],
      dtype='object')

In [22]:
# Use the describe function to gather some basic statistics
client_data_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number      subtotal  \
count  54639.000000  54639.000000  5.463900e+04  54639.000000  5.463900e+04   
mean      99.446073      5.004116  5.702646e+02      2.979667  7.710286e+04   
std      133.164267      5.326599  1.879552e+04      2.436320  2.881926e+06   
min        0.010000      0.000000  0.000000e+00      0.000000  0.000000e+00   
25%       14.840000      1.450000  3.200000e+01      1.000000  1.132950e+03   
50%       49.890000      3.240000  6.800000e+01      3.000000  4.629440e+03   
75%      125.570000      6.890000  1.700000e+02      5.000000  1.736913e+04   
max      846.270000     46.430000  3.958244e+06      9.000000  5.844743e+08   

       shipping_cost  shipping_weight    line_total     line_cost  \
count   5.463900e+04     5.463900e+04  5.463900e+04  5.463900e+04   
mean    1.742510e+04     2.487416e+03  1.032718e+05  7.387972e+04   
std     3.891595e+05     5.559430e+04  3.286505e+06  2.133294e+06   
min     0.000000e+00     0.000000e+00  0.000000e+00  0.000000e+00   
25%     4.966500e+02     7.033000e+01  2.887560e+03  2.206360e+03   
50%     1.577800e+03     2.254000e+02  8.396490e+03  6.266040e+03   
75%     5.029080e+03     7.184400e+02  2.645604e+04  1.946590e+04   
max     6.306339e+07     9.009056e+06  6.412625e+08  3.987535e+08   

        line_profit  
count  5.463900e+04  
mean   2.939207e+04  
std    1.195291e+06  
min    0.000000e+00  
25%    5.597750e+02  
50%    1.894840e+03  
75%    6.674210e+03  
max    2.425090e+08

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data
print(client_data_df.dtypes)
client_data_df.describe(include='all')

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object


first     last        job       phone                     email  \
count         54639    54639      54639       54639                     54639   
unique          360      477        498         999                       999   
top     Christopher  Johnson  Osteopath  5016915923  jessica.reyes@grimes.net   
freq           1090     1164        631         220                       220   
mean            NaN      NaN        NaN         NaN                       NaN   
std             NaN      NaN        NaN         NaN                       NaN   
min             NaN      NaN        NaN         NaN                       NaN   
25%             NaN      NaN        NaN         NaN                       NaN   
50%             NaN      NaN        NaN         NaN                       NaN   
75%             NaN      NaN        NaN         NaN                       NaN   
max             NaN      NaN        NaN         NaN                       NaN   

           client_id      order_id  order_date    order_week    order_year  \
count   54639.000000  5.463900e+04       54639  54639.000000  54639.000000   
unique           NaN           NaN         150           NaN           NaN   
top              NaN           NaN  2023-04-19           NaN           NaN   
freq             NaN           NaN         538           NaN           NaN   
mean    54837.869416  5.470190e+06         NaN     11.359139   2022.993064   
std     25487.438231  2.599807e+06         NaN      7.023499      0.082997   
min     10033.000000  1.000886e+06         NaN      1.000000   2022.000000   
25%     33593.000000  3.196372e+06         NaN      6.000000   2023.000000   
50%     53305.000000  5.496966e+06         NaN     11.000000   2023.000000   
75%     78498.000000  7.733869e+06         NaN     17.000000   2023.000000   
max     99984.000000  9.998480e+06         NaN     52.000000   2023.000000   

               item_id     category        subcategory    unit_price  \
count            54639        54639              54639  54639.000000   
unique             500            5                 18           NaN   
top     MAZ00314-80-8M  consumables  bathroom supplies           NaN   
freq               148        23538               6424           NaN   
mean               NaN          NaN                NaN    136.267207   
std                NaN          NaN                NaN    183.873135   
min                NaN          NaN                NaN      0.010000   
25%                NaN          NaN                NaN     20.800000   
50%                NaN          NaN                NaN     68.310000   
75%                NaN          NaN                NaN    173.160000   
max                NaN          NaN                NaN   1396.230000   

           unit_cost   unit_weight           qty   line_number  
count   54639.000000  54639.000000  5.463900e+04  54639.000000  
unique           NaN           NaN           NaN           NaN  
top              NaN           NaN           NaN           NaN  
freq             NaN           NaN           NaN           NaN  
mean       99.446073      5.004116  5.702646e+02      2.979667  
std       133.164267      5.326599  1.879552e+04      2.436320  
min         0.010000      0.000000  0.000000e+00      0.000000  
25%        14.840000      1.450000  3.200000e+01      1.000000  
50%        49.890000      3.240000  6.800000e+01      3.000000  
75%       125.570000      6.890000  1.700000e+02      5.000000  
max       846.270000     46.430000  3.958244e+06      9.000000

In [26]:
# What three item categories had the most entries?
print (client_data_df['category'].value_counts().head(3).to_string(header=False,dtype=False),'\n')
client_data_df['category'].value_counts().head(3)

consumables    23538
furniture      11915
software        8400 



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries,
# which subcategory had the most entries?
print (client_data_df.groupby(['category','subcategory']).size().nlargest(1).to_string(header=False))


consumables  bathroom supplies    6424


In [8]:
# Which five clients had the most entries in the data? check with just last name and then with first name
 
print (client_data_df.groupby(['first','last','client_id']).size().nlargest(5).to_string(header=False,dtype=False))

Jessica    Reyes    33615        220
Angela     Everett  66037        211
Bryan      Myers    46820        209
Alexandra  Young    38378        207
Kendra     Garrett  24741        207


In [9]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = client_data_df.groupby(['client_id']).size().nlargest(5).index.tolist()
top_client_ids



[33615, 66037, 46820, 24741, 38378]

In [10]:
# How many total units (the qty column) did the
# client with the most entries order order?
client_data_df[client_data_df['client_id'] == top_client_ids[0]]['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
client_data_df['subtotal'] = client_data_df.apply((lambda x: x['unit_price']*x['qty']),axis=1)
print (client_data_df[['unit_price','qty','subtotal']].head(2).to_string(index=False),'\n')


 unit_price  qty  subtotal
    1096.80  105 115164.00
      24.95   21    523.95 



In [31]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
client_data_df['shipping_weight'] = client_data_df.apply(lambda row: row['unit_weight'] * row['qty'],axis=1)
client_data_df['shipping_cost']   = client_data_df.apply(lambda row: 7.00 * row['shipping_weight'] 
                                    if row['unit_weight'] * row['qty'] > 50
                                    else 10.00 * row['unit_weight'] * row['qty'],axis=1)
print(client_data_df[['unit_price','unit_weight', 'qty', 'shipping_weight','shipping_cost']].head(3).to_string(index=False))

 unit_price  unit_weight  qty  shipping_weight  shipping_cost
    1096.80         7.50  105           787.50        5512.50
      24.95         1.49   21            31.29         312.90
      13.52         1.68   39            65.52         458.64


In [13]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

sales_tax = 1.0925
client_data_df['line_total'] = round(client_data_df.apply(lambda row: ((row['subtotal']) + row['shipping_cost'])*sales_tax,axis=1),2)
print(client_data_df[['qty', 'subtotal','shipping_cost','line_total']].head(3))

   qty   subtotal  shipping_cost  line_total
0  105  115164.00        5512.50   131839.08
1   21     523.95         312.90      914.26
2   39     527.28         458.64     1077.12


In [33]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

client_data_df['line_cost'] = client_data_df.apply(lambda row: row['unit_cost'] * row['qty'] + row['shipping_cost'], axis=1)
client_data_df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

    subtotal  shipping_cost  shipping_weight  line_total  line_cost  \
0  115164.00        5512.50           787.50   131839.08   85597.05   
1     523.95         312.90            31.29      914.26     629.79   
2     527.28         458.64            65.52     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

In [15]:
# Create a column for the profit of
# each line using line cost and line price
client_data_df['line_profit'] = client_data_df.apply(lambda row: row['line_total'] -row['line_cost'] , axis=1)

print(pd.concat([client_data_df['order_id'], client_data_df[['line_cost', 'line_total', 'line_profit']]
                 .map(format_currency)], axis=1).head().to_string(index=False))


 order_id  line_cost  line_total line_profit
  8953482 $85,597.05 $131,839.08  $46,242.03
  8069089    $629.79     $914.26     $284.47
  1902144    $765.18   $1,077.12     $311.94
  9031802  $1,077.35   $1,543.57     $466.22
  1322274  $8,663.60  $11,364.40   $2,700.80


## Part 3: Confirm your work
You have email receipts showing that the line_total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [34]:
# Check your work using the totals above
for order_id in [2742071, 2173913, 6128929]:
    # Calculate the sum
    customer_order_total = client_data_df.loc[client_data_df['order_id'] == order_id]['line_total'].sum()
    # format and print results
    print (f'Order {order_id}  Total ${customer_order_total:.2f}')

Order 2742071  Total $152811.89
Order 2173913  Total $162388.71
Order 6128929  Total $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

How much did each of the top 5 clients by quantity spend? 
Check your work from Part 1 for client ids.
[33615, 66037, 46820, 24741, 38378]

In [17]:
top_client_ids = client_data_df.groupby(['client_id']).size().nlargest(5).index.to_list()
for client_id in top_client_ids:
    # Calculate the sum
    customer_order_total = client_data_df.loc[client_data_df['client_id'] == client_id]['line_total'].sum()
    # format and print results
    print (f'Client {client_id}\tTotal\t${customer_order_total:.2f}')

Client 33615	Total	$8377308.52
Client 66037	Total	$10259514.79
Client 46820	Total	$9743794.36
Client 24741	Total	$82268892.02
Client 38378	Total	$12906550.87


Create a summary DataFrame showing the totals for the
for the top 5 clients with the following information:
total units purchased, total shipping price,
total revenue, and total profit. Sort by total profit.

In [35]:
# Create an empty summary DataFrame with the specified columns and data types
summary_df = pd.DataFrame(columns=['client_id', 'total_qty', 'total_shipping', 'total_revenue', 'total_cost', 'total_profit'])

# Iterate over the top 5 client IDs
for client_id in top_client_ids:
    # Calculate totals for the current client
    total_qty      = client_data_df.loc[client_data_df['client_id'] == client_id]['qty'].sum()
    total_shipping = client_data_df.loc[client_data_df['client_id'] == client_id]['shipping_cost'].sum()
    total_revenue  = client_data_df.loc[client_data_df['client_id'] == client_id]['line_total'].sum()
    total_cost     = client_data_df.loc[client_data_df['client_id'] == client_id]['line_cost'].sum()
    total_profit   = client_data_df.loc[client_data_df['client_id'] == client_id]['line_profit'].sum()
    
    # Create a new DataFrame for the current client
    new_row_df = pd.DataFrame({'client_id'     : [client_id],
                               'total_qty'     : [total_qty],
                               'total_shipping': [total_shipping],
                               'total_revenue' : [total_revenue],
                               'total_cost'    : [total_cost],
                               'total_profit'  : [total_profit]})
    
    # Concatenate the new DataFrame with the summary DataFrame
    summary_df = pd.concat([summary_df, new_row_df], ignore_index=True, axis=0)

# Sort the summary DataFrame by total_profit in descending order
summary_df = summary_df #.sort_values(by='total_profit', ascending=False)

# Print the sorted summary DataFrame
summary_df.sort_values(by="total_profit",ascending=False)

C:\Users\framk\AppData\Local\Temp\ipykernel_12684\2827033578.py:22: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  summary_df = pd.concat([summary_df, new_row_df], ignore_index=True, axis=0)


client_id total_qty  total_shipping  total_revenue   total_cost  \
3     24741    239862      5126448.37    82268892.02  45688899.71   
4     38378     73667      3429455.40    12906550.87   9634720.98   
1     66037     43018      1395151.85    10259514.79   7004482.98   
2     46820     75768      1601448.84     9743794.36   7007191.64   
0     33615     64313      1828984.89     8377308.52   6175313.91   

   total_profit  
3   36579992.31  
4    3271829.89  
1    3255031.81  
2    2736602.72  
0    2201994.61

In [19]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
summary_df=summary_df.rename(columns=
    {'client_id':'Client ID', 
     'total_qty':'Units', 
     'total_shipping':'Shipping', 
     'total_revenue':'Total Revenue',
     'total_cost':'Total Cost', 
     'total_profit':'Total Profit'
})
# summary_df[['Client ID','Units','Shipping','Total Revenue','Total Cost','Total Profit']]
display_df = summary_df.assign(**{
    'Shipping'     : lambda x: ('$' + (x['Shipping']      / 1_000_000).round(2).astype(str) + 'M'),
    'Total Revenue': lambda x: ('$' + (x['Total Revenue'] / 1_000_000).round(2).astype(str) + 'M'),
    'Total Cost'   : lambda x: ('$' + (x['Total Cost']    / 1_000_000).round(2).astype(str) + 'M'),
    'Total Profit' : lambda x: ('$' + (x['Total Profit']  / 1_000_000).round(2).astype(str) + 'M')
})
display_df

Client ID   Units Shipping Total Revenue Total Cost Total Profit
0     33615   64313   $1.83M        $8.38M     $6.18M        $2.2M
1     66037   43018    $1.4M       $10.26M      $7.0M       $3.26M
2     46820   75768    $1.6M        $9.74M     $7.01M       $2.74M
3     24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
4     38378   73667   $3.43M       $12.91M     $9.63M       $3.27M

In [20]:
# Sort the updated data by "Total Profit" form highest to lowest
display_df.sort_values(by="Total Profit",ascending=False)

Client ID   Units Shipping Total Revenue Total Cost Total Profit
3     24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
4     38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1     66037   43018    $1.4M       $10.26M      $7.0M       $3.26M
2     46820   75768    $1.6M        $9.74M     $7.01M       $2.74M
0     33615   64313   $1.83M        $8.38M     $6.18M        $2.2M